clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java

```diff
@@ -237,4 +253,64 @@ public void updateAssignment(ConsumerGroupHeartbeatResponseData.Assignment assig
         }
         maybeTransitionToStable();
     }
+
+    /**
+     * Performs the reconciliation between the current topic assignment and the {@link Assignment} provided in the
+     * heartbeat response. Per the KIP-848 protocol, we perform the following:
+     *
+     * <ol>
+     *     <li>Revoke partitions, if any</li>
+     *     <li>Heartbeat to acknowledge revoked partitions</li>
+     *     <li>Assign partitions, if any</li>
+     *     <li>Heartbeat to acknowledge assigned partitions</li>
+     * </ol>
+     *
+     * TODO: What are the the state changes here?
+     * TODO: Is SubscriptionState sufficient to build the next heartbeat request?
+     * TODO: Where do we need to call ConsumerRebalanceListener.onPartitionsLost()?
+     */
+    void reconcile() {
+        transitionTo(MemberState.RECONCILING);
+
+        Timer remainingAssignmentTime = time.timer(10000);
+
+        // First, we need to determine if any partitions need to be revoked.
+        {
+            ReconciliationResult result = reconciler.revoke(targetAssignment, remainingAssignmentTime);
+            remainingAssignmentTime.update();
+
+            if (result == ReconciliationResult.COMPLETED) {
+                // If we've revoked one or more partitions, we need to send an acknowledgement request ASAP to
+                // let the coordinator know that they've been removed locally.
+                return;
+            } else if (result == ReconciliationResult.EXPIRED) {
+                // TODO: what do we do here?
```

Comment:
This is the description from kip-848 
```
The chosen member is expected to complete the assignment process within the rebalance timeout. The time on the coordinator side starts ticking when the member is notified. If the process is not completed within the rebalance timeout, the group coordinator picks up another member to run the assignment. Note that the previous chosen member is not fenced here because the fencing is only done based on the session.
```

It seems that you should send a heartbeat to the GC and get a fenced exception, which results in revoking all partitions.

@dajac can provide a more accurate answer here.

Reply:
I may be missing something, but I don't quite get why we need to time this on the client side. It is the server the one that handles the rebalanceTimeouts. 

The server will start the timer when sending an assignment/revocation to a member, and if it does not receive the ack in time, it will fence the member. From the client point of view, when it completes processing the assignment/revocation, it would notice it has been fenced (state is FENCED), and it won't send the ack to the assignment (because it is not part of the group anymore). It would have to invoke the onPartitionsLost and rejoin the group I expect.

Reply:
Should I remove the `Timer` and expose an `onFenced()` method that the heartbeat invokes? Then the `onFenced()` implementation would call `onPartitionsLost()`, right?

Reply:
I would say we should aim for keeping only state updates and member info in the MembershipMgr, not the logic for triggering the `onPartitionsLost`. It could be maybe the assignment reconciler that, when it finishes processing, triggers a HB or a partitionsLost depending on the state (if state is still reconciling and it finished processing ok we just need to HB, if state is fenced then we don't HB. We trigger onPartitionsLost and then trigger HB to rejoin). 
Seems all like interaction between the AssginemntReconciler and the HB request, only checking the state to get the current state and determine what to do (and eventually updating the state with the outcome)